[BB2-896] Responsive Design NodeJS React Sample (#7)
CMSgov/bluebutton-sample-client-nodejs-react: ErrorResponses.md

# Blue Button 2.0 API Error Responses

## Overview

> This document serves as a supplementary to Blue Button 2.0 API Developer Documents, it gives more details on most commonly error responses and recommended error handling on the client side, e.g. guide lines on retrying the requests on certain errors.

Reviewer: ```suggestion
This document serves as a supplementary to Blue Button 2.0 API Developer Documents. It gives more details on the most common error responses and how to properly handle them.

```

| 500<br>SERVER ERROR | /v[12]/fhir/.* | The root cause of the 500 error, some times, is indicated by the error message, <br>the app can choose to retry the failed request depend on the nature of the root cause, <br>examples that might be retriable are those related to network down (temporarily):<br>Example:<br>ConnectionError at /mymedicare/login<br>HTTPSConnectionPool(host='test.accounts.cms.gov', port=443): <br>Max retries exceeded with url: /health (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x7f46599dafd0>: <br>Failed to establish a new connection: [Errno -2] Name or service not known')) | Heuristic on Retry | App can choose to retry on some of the 500 errors as shown by the example, this is a heuristic approach. |

### Retry

> App can choose to retry the requests on certain 500 errors heuristically depending on the root cause as indicated by the error messages, e.g. backend server is unavailable due to temporarily network down etc., the app can choose from various retry patterns that best fits the use case (e.g. exponential back off).

Reviewer: ```suggestion

```

> Auto retrying on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended, but with sensible retry settings.

Reviewer: ```suggestion
Auto retrying (with sensible retry settings) on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended.

```

> Due to the involvement of end user (beneficiary), auto retrying on requests in an authorization flow does not seem practical, hence not recommended.

Reviewer: ```suggestion
Due to the involvement of the end user (beneficiary), auto retrying requests in the authorization flow are not recommended.

```

# Blue Button 2.0 API Error Responses

## Overview

This document serves as a supplementary to Blue Button 2.0 API Developer Documents, it gives more details on most commonly error responses and recommended error handling on the client side, e.g. guide lines on retrying the requests on certain errors.

## Error Responses and Client Actions

### Authorization Requests

> | Status Code    | End Point URL   | Error Message              | Action            | Comments                                               |

Reviewer: I'm trying to decide if this is just too much information. Should we just highlight which type of requests they could retry on? Everything else feels a bit too much for a sample app. Basically it's a lot of text that essentially sums up to: "Don't do anything".

Author: though the purpose is to give more bb2 specifics, since HTTP response codes are already defined.

Can reduce the verbose-ness a bit.

Reviewer: Want to take an attempt at trimming it down? Really focus on what they would need to know in order to successfully build an application against BB2.

Author: those are frequently used error responses among the full set http status codes, and it contains bb2 specific info, so there are added values, or let me know what should be trimmed? don't mind remove the whole mark down - remember it was added based on some feedback-:

Reviewer: My only concern is that this table is a lot of text. I feel like most people wont read it all, and then they are missing out on important information. Were we to trim it down to just when they need to do something (ie retry) then they might catch it. If you feel strongly we can leave it though. Just clean up the typing issues I pointed out and we should be able to get this approved 👍

Author: for the enforce of typing, for those places that the type of the var or param is obvious, explicit type is given, but some places e.g. the err var of axios try catch, a single type could not cover all situations, e.g. the error thrown from axios end point call could be AxiosError but it could be something else, there are discussion on this from internet..., also ran across some people's code to handle the multiple error types - but it will pull in more code, may be we could log a backlog for "making node js react sample strong typed" as future improvement? thanks for brought this up.
